text/0000-build-observability.md

```diff
@@ -0,0 +1,358 @@
+                    "stringValue": "heroku/nodejs-engine"
+                  }
+                }
+              ],
+              "droppedEventsCount": 1
+            }
+          ]
+        }
+      ]
+    }
+  ]
+}
+```
+
+### Lifetime
+
+Telemetry files may be written at any point during the build, so that they
+are persisted in cases of failures to detect, failures to build, process
+terminations, or crashes. The `jsonl` format allows telemetry libraries to
+safely append additional json objects to the end of a telemetry file, so
+telemetry data can be flushed to the file frequently. Telemetry files should
+not be truncated or deleted so that telemetry processing by a platform can
+happen during or after a build. Telemetry files should not be included in the
+build result, as they are not relevant, and would likely negatively impact
+image size and reproduceability.
+
+### Access
+
+The telemetry files should be readable so that they may be analyzed by
+the user and/or platform. However, they should be write protected
+to prevent malicious buildpacks from injecting tracing data into other
+buildpack or lifecycle telemetry files.
+
+
+### Consumption
+
+This RFC leaves the consumption of telemetry files to the platform operator.
+Platform operators choosing to use these metrics need to read them either during
+or after the build. This can be done using existing OpenTelemetry libraries.
+Platform operators may choose to optionally enrich or modify the tracing data
+as they see fit (with data like `instance_id` or `build_id`). Platform
+operators will likely want to export this data to an OpenTelemetry backend for
+persistence and analysis, and again, this may be done with existing
+OpenTelemetry libraries.
+
+### Viewing and Analyzing
+
+Once the lifecycle and buildpack traces are exported to an OpenTelemetry
+backend, platform operators should be able to (depending on the features of the
+backend):
+
+- View the complete trace for a build
+- View or query attributes attached to spans (e.g. `buildpack_id`,
+  `nodejs_version`)
+- View or query span durations
+- View or query error types and/or messages
+- and more
+
+# Migration
+[migration]: #migration
+
+No migration neccessary, this is net-new functionality with no backwards
+compatibilty concers.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+### Privacy Concerns
+
+This RFC outlines functionality that could be percieved as user tracking. To
+help remediate those concerns, these are some factors to remember about this
+design:
+
+1) This functionality is opt-in. `lifecycle` and `pack` will not emit telemetry
+  data unless the `--telemetry` flag is used.
+2) This functionality emits telemetry data only to the build file system. For
+  `pack` users, the telemetry files are stored in docker volumes on the local
+  machine. Neither `pack` nor `lifecycle` will "phone home" with telemety data.
+3) Neither `pack` nor `lifecycle` collect user-identifiable data (no emails,
```

Comment:
Should image/registry names be considered sensitive?

Reply:
Hmm - tough question. Yeah, I suppose some folks would consider the resulting image name/registry sensitive. I'll update.